libdnf/module/module_package.cpp

```diff
@@ -0,0 +1,280 @@
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/module/module_package.hpp"
+
+#include <modulemd-2.0/modulemd-module-stream.h>
+#include <modulemd-2.0/modulemd-profile.h>
+#include <modulemd-2.0/modulemd.h>
+
+#include <algorithm>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace libdnf::module {
+
+
+const char * ModulePackage::get_name_cstr() const {
+    return modulemd_module_stream_get_module_name(md_stream);
+}
+
+
+std::string ModulePackage::get_name() const {
+    auto name = modulemd_module_stream_get_module_name(md_stream);
+    return name ? name : "";
+}
+
+
+const char * ModulePackage::get_stream_cstr() const {
+    return modulemd_module_stream_get_stream_name(md_stream);
+}
+
+
+std::string ModulePackage::get_stream() const {
+    auto name = modulemd_module_stream_get_stream_name(md_stream);
+    return name ? name : "";
+}
+
+
+long long ModulePackage::get_version_num() const {
+    return (long long)modulemd_module_stream_get_version(md_stream);
+}
+
+
+std::string ModulePackage::get_version() const {
+    return std::to_string(modulemd_module_stream_get_version(md_stream));
+}
+
+
+const char * ModulePackage::get_context_cstr() const {
+    return modulemd_module_stream_get_context(md_stream);
+}
+
+
+std::string ModulePackage::get_context() const {
+    auto context = modulemd_module_stream_get_context(md_stream);
+    return context ? context : "";
+}
+
+
+const char * ModulePackage::get_arch_cstr() const {
+    return modulemd_module_stream_get_arch(md_stream);
+}
+
+
+std::string ModulePackage::get_arch() const {
+    auto arch = modulemd_module_stream_get_arch(md_stream);
+    return arch ? arch : "";
+}
+
+
+std::string ModulePackage::get_name_stream(ModulemdModuleStream * md_stream) {
+    std::ostringstream ss;
+    auto name = modulemd_module_stream_get_module_name(md_stream);
+    auto stream = modulemd_module_stream_get_stream_name(md_stream);
+    ss << (name ? name : "") << ":" << (stream ? stream : "");
+    return ss.str();
+}
+
+
+std::string ModulePackage::get_name_stream_version() const {
+    std::ostringstream ss;
+    ss << get_name_stream() << ":" << get_version();
+    return ss.str();
+}
+
+
+std::string ModulePackage::get_full_identifier() const {
+    std::ostringstream ss;
+    ss << get_name() << ":" << get_stream() << ":" << get_version() << ":" << get_context() << ":" << get_arch();
```

Comment:
It is quite slow approach how to merge strings together.

Reply:
What about to compare the performance with following code?
```
auto result = get_name();
result.append(":")
result.append(get_stream_cstr())
result.append(":")
result.append(get_version())
result.append(":")
result.append(get_context_cstr())
```

Reply:
Here the best probably would be `libdnf::utils::sformat()` (at least from code cleanliness perspective, but performance comparison would be interesting)